database/src/table.rs

```diff
@@ -1,181 +1,49 @@
-//! ### Table module
-//! This module contains the definition of the [`Table`] and [`DupTable`] trait, and the actual tables used in the database.
-//! [`DupTable`] are just a trait used to define that they support DUPSORT|DUPFIXED operation (as of now we don't know the equivalent for HSE).
-//! All tables are defined with docs explaining its purpose, what types are the key and data.
-//! For more details please look at Cuprate's book : <link to cuprate book>
-
-use crate::{
-    encoding::Compat,
-    types::{
-        /*OutTx,*/ AltBlock, BlockMetadata, /*RctOutkey,*/ OutputMetadata,
-        TransactionPruned, TxIndex, /*OutAmountIdx,*/ /*KeyImage,*/ TxOutputIdx,
-    },
-};
-use bincode::{de::Decode, enc::Encode};
-use monero::{blockdata::transaction::KeyImage, Block, Hash};
-
-/// A trait implementing a table interaction for the database. It is implemented to an empty struct to specify the name and table's associated types. These associated
-/// types are used to simplify deserialization process.
-pub trait Table: Send + Sync + 'static + Clone {
-    // name of the table
-    const TABLE_NAME: &'static str;
-
-    // Definition of a key & value types of the database
-    type Key: Encode + Decode;
-    type Value: Encode + Decode;
-}
-
-/// A trait implementing a table with duplicated data support.
-pub trait DupTable: Table {
-    // Subkey of the table (prefix of the data)
-    type SubKey: Encode + Decode;
-}
-
-/// This declarative macro declare a new empty struct and impl the specified name, and corresponding types.
-macro_rules! impl_table {
-	( $(#[$docs:meta])* $table:ident , $key:ty , $value:ty ) => {
-        #[derive(Clone)]
-		$(#[$docs])*
-		pub(crate) struct $table;
-
-   		impl Table for $table {
-	 		const TABLE_NAME: &'static str = stringify!($table);
-        	type Key = $key;
-            type Value = $value;
+//! Database table abstraction; `trait Table`.
+
+//---------------------------------------------------------------------------------------------------- Import
+use crate::pod::Pod;
+
+//---------------------------------------------------------------------------------------------------- Table
+/// Database table metadata.
+///
+/// Purely compile time information for database tables.
+/// Not really an accurate name for `K/V` database but
+/// this represents the metadata of a `K/V` storing object.
+pub trait Table {
```

Comment:
This table should probably contain more information about the data it's storing so the backend can optimize for certain tables.

Currently I think the only thing that's missing that we would be able to optimize for is:
`const CONSTANT_SIZE: bool`
 
for if the table's values are all the same size.